plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XProjectManager.java

```diff
@@ -133,6 +147,62 @@ public XBuildResult doIncrementalBuild(Set<URI> dirtyFiles, Set<URI> deletedFile
 		return result;
 	}
 
+	/** Deletes the contents of the output directory */
+	public void doClean(CancelIndicator cancelIndicator) {
+		projectStateHolder.deletePersistenceFile(projectConfig);
+
+		if (projectConfig instanceof N4JSProjectConfig) {
+			// TODO: merge N4JSProjectConfig#canClean() to IProjectConfig
+			N4JSProjectConfig n4pc = (N4JSProjectConfig) projectConfig;
+			if (!n4pc.canClean()) {
+				return;
+			}
+		}
+
+		XBuildRequest request = buildRequestFactory.getBuildRequest();
+		for (File outputDirectory : getOutputDirectories()) {
+			File[] childFildes = outputDirectory.listFiles();
+			if (childFildes != null) {
+				for (int i = 0; i < childFildes.length; i++) {
+					deleteFileOrFolder(request, childFildes[i]);
+				}
+			}
+
+			operationCanceledManager.checkCanceled(cancelIndicator);
+		}
+	}
+
+	/** @return list of output directories of this project */
+	public List<File> getOutputDirectories() {
+		List<File> outputDirs = new ArrayList<>();
+		Set<OutputConfiguration> outputConfigurations = outputConfigProvider.getOutputConfigurations(resourceSet);
+		URI projectBaseUri = projectConfig.getPath();
+		for (OutputConfiguration outputConf : outputConfigurations) {
+			for (String outputDirName : outputConf.getOutputDirectories()) {
+				URI outputUri = projectBaseUri.appendSegment(outputDirName);
+				File outputDirectory = new File(outputUri.toFileString());
+				outputDirs.add(outputDirectory);
+			}
+		}
+		return outputDirs;
+	}
+
+	/** Deletes the given file recursively */
+	protected void deleteFileOrFolder(XBuildRequest request, File file) {
```

Comment:
Have you checked if you can use our `FileDeleter` utility class for that?

Reply:
I wasn't aware of that, thanks. I also tried and made all changes until I realised that `FileDeleter` is part of n4js packages. Inside `n4js.ide.xtext.*` packages, I try not to reference classes outside of these packages.